# Analisis Ekstraksi Fitur Deret Waktu & K-Means Clustering Polutan Udara (CO, NO2, SO2)

Notebook ini berisi analisis komprehensif eksplorasi data deret waktu (*time series*), ekstraksi fitur berbasis **TSFEL**, reduksi dimensi menggunakan **PCA (*Principal Component Analysis*)**, serta penentuan jumlah klaster optimal pada algoritma **K-Means Clustering** untuk 3 jenis polutan udara utama: **Karbon Monoksida (CO)**, **Nitrogen Dioksida (NO2)**, dan **Sulfur Dioksida (SO2)**.

Analisis ini dikerjakan secara terintegrasi menggunakan dua pendekatan:
1. **Pendekatan Tools**: Menggunakan **KNIME Analytics Platform** dengan alur *PostgreSQL Connector &rarr; DB Table Selector &rarr; DB Reader &rarr; PCA &rarr; k-Means &rarr; Scatter Plot*.
2. **Pendekatan Code**: Menggunakan **Python (Scikit-Learn, Pandas, Matplotlib, Seaborn)** di dalam Jupyter Notebook ini.

---

## 1. Konsep Dasar Statistik & Deskripsi 68 Fitur TSFEL

Data polutan harian yang diperoleh dari observasi satelit Sentinel-5P diekstraksi menggunakan pustaka **TSFEL (*Time Series Feature Extraction Library*)**. Sebanyak **68 fitur numerik** diekstraksi dari setiap sinyal deret waktu dan dikelompokkan ke dalam 4 domain utama:

### A. Domain Statistik (16 Fitur)
Mengukur tendensi sentral, dispersi, dan bentuk sebaran data konsentrasi polutan:
- **`calc_mean`**: Rata-rata aritmatika konsentrasi polutan selama periode observasi harian.
- **`calc_std` & `calc_var`**: Standar deviasi dan variansi, mengukur tingkat fluktuasi data terhadap nilai rata-ratanya.
- **`calc_median`**: Nilai tengah persentil 50%, ukuran pemusatan yang tahan terhadap nilai ekstrem (*outliers*).
- **`calc_min` & `calc_max`**: Konsentrasi polutan terendah dan tertinggi yang tercatat.
- **`interq_range` (IQR)**: Rentang interkuartil ($Q_3 - Q_1$), mengukur sebaran 50% data di tengah distribusi.
- **`skewness`**: Derajat kemiringan distribusi. Skewness positif mengindikasikan bahwa sebagian besar hari konsentrasi polutan rendah dengan sesekali lonjakan polusi ekstrem.
- **`kurtosis`**: Keruncingan distribusi. Kurtosis tinggi menandakan keberadaan pencilan ekstrem (*heavy tails*).
- **`hist_mode`**: Modus data (nilai konsentrasi yang paling sering muncul).
- **`mean_abs_deviation` & `median_abs_deviation` (MAD)**: Rata-rata dan median selisih mutlak titik data terhadap titik pusatnya.
- **`ecdf`, `ecdf_percentile`, `ecdf_percentile_count`, `ecdf_slope`**: Parameter fungsi distribusi kumulatif empiris untuk melihat laju akumulasi probabilitas konsentrasi polutan.

### B. Domain Temporal / Waktu (18 Fitur)
Mengukur sifat dinamis, laju perubahan antar-waktu, dan total akumulasi energi sinyal:
- **`abs_energy`**: Total energi sinyal ($\sum x_i^2$), merepresentasikan akumulasi beban pencemaran sepanjang tahun.
- **`auc` (Area Under Curve)**: Luas area di bawah kurva konsentrasi polutan harian.
- **`autocorr`**: Autokorelasi sinyal, mengukur ketergantungan konsentrasi hari ini dengan hari-hari sebelumnya (persistensi temporal).
- **`average_power`**: Rata-rata daya sinyal per satuan waktu.
- **`calc_centroid`**: Titik berat waktu sinyal, mengindikasikan kapan puncak pencemaran cenderung terkonsentrasi dalam kalender tahunan.
- **`distance`**: Total jarak lintasan sinyal (panjang kurva fluktuasi harian).
- **`mean_diff`, `mean_abs_diff`, `median_diff`, `median_abs_diff`, `sum_abs_diff`**: Rata-rata, median, dan akumulasi perbedaan konsentrasi antar-hari berturut-turut ($x_{t+1} - x_t$).
- **`slope`**: Kemiringan garis tren linier, menunjukkan tren jangka panjang apakah emisi cenderung meningkat atau menurun.
- **`pk_pk_distance`**: Rentang puncak-ke-lembah ($Max - Min$).
- **`zero_cross`, `positive_turning`, `negative_turning`, `neighbourhood_peaks`**: Jumlah perlintasan batas dasar serta frekuensi titik belok puncak (*local maxima*) dan lembah (*local minima*).
- **`rms` (Root Mean Square)**: Nilai akar kuadrat rata-rata dari kuadrat sinyal.

### C. Domain Frekuensi / Spektral (26 Fitur)
Mentransformasikan sinyal waktu ke domain frekuensi (melalui *Fast Fourier Transform* / *Wavelet*) untuk menangkap pola siklus berkala (misalnya siklus musiman atau mingguan):
- **`fundamental_frequency`, `max_frequency`, `median_frequency`**: Frekuensi dasar dan komponen frekuensi dengan daya spektral dominan.
- **`max_power_spectrum`, `power_bandwidth`**: Magnitudo daya puncak dan lebar pita energi sinyal.
- **`spectral_centroid`, `spectral_spread`, `spectral_skewness`, `spectral_kurtosis`**: Momen spektral yang menggambarkan pusat gravitasi frekuensi, sebaran, dan keruncingan spektrum.
- **`spectral_slope`, `spectral_decrease`, `spectral_roll_off`, `spectral_roll_on`**: Karakteristik penurunan energi pada frekuensi tinggi.
- **`spectral_entropy`**: Tingkat keacakan atau kompleksitas spektrum frekuensi polutan.
- **`wavelet_energy`, `wavelet_entropy`, `wavelet_abs_mean`, `wavelet_std`, `wavelet_var`**: Energi dan variabilitas hasil dekomposisi gelombang (*wavelet*).
- **`mfcc`, `lpcc`, `human_range_energy`**: Representasi koefisien spektral lanjutan.

### D. Domain Fraktal & Kompleksitas (8 Fitur)
Mengukur ketidakteraturan, sifat kekacauan (*chaos*), dan memori jangka panjang deret waktu:
- **`entropy` (Shannon Entropy) & `mse` (Multiscale Entropy)**: Derajat keacakan informasi sinyal pada skala waktu tunggal maupun jamak.
- **`hurst_exponent`**: Menguji sifat persistensi data ($H > 0.5$ mengindikasikan tren jangka panjang; $H < 0.5$ bersifat *mean-reverting*).
- **`dfa` (Detrended Fluctuation Analysis)**: Korelasi rentang panjang pada deret waktu non-stasioner.
- **`higuchi_fractal_dimension` & `petrosian_fractal_dimension`**: Dimensi fraktal yang mengukur kekasaran geometris sinyal.
- **`maximum_fractal_length` & `lempel_ziv`**: Kompleksitas munculnya pola-pola baru dalam sinyal.

---

### Contoh Perhitungan Manual vs Pembuktian TSFEL
Misalkan terdapat sampel deret waktu konsentrasi polutan selama 4 hari:
$$X = [2.0, \, 4.0, \, 6.0, \, 8.0]$$

1. **Perhitungan Manual Rata-rata (`calc_mean`):**
   $$\mu = \frac{\sum_{i=1}^{N} x_i}{N} = \frac{2.0 + 4.0 + 6.0 + 8.0}{4} = \frac{20.0}{4} = 5.0$$

2. **Perhitungan Manual Rentang Puncak-ke-Lembah (`pk_pk_distance`):**
   $$\text{Peak-to-Peak} = \max(X) - \min(X) = 8.0 - 2.0 = 6.0$$

3. **Perhitungan Manual Root Mean Square (`rms`):**
   $$\text{RMS} = \sqrt{\frac{1}{N} \sum_{i=1}^{N} x_i^2} = \sqrt{\frac{2^2 + 4^2 + 6^2 + 8^2}{4}} = \sqrt{\frac{4 + 16 + 36 + 64}{4}} = \sqrt{\frac{120}{4}} = \sqrt{30} \approx 5.4772$$

4. **Perhitungan Manual Total Energi (`abs_energy`):**
   $$\text{Energy} = \sum_{i=1}^{N} x_i^2 = 4 + 16 + 36 + 64 = 120.0$$

*Hasil perhitungan manual di atas telah divalidasi dan terbukti identik 100% dengan nilai yang dihitung oleh fungsi-fungsi TSFEL.*

---

## 2. Struktur Dataset Ekstraksi Fitur untuk 1 Kelas
Dataset yang dianalisis merupakan gabungan data hasil ekstraksi fitur deret waktu dari **1 kelas** yang terdiri atas **37 mahasiswa / wilayah kajian**:
- **Jumlah Sampel (Baris):** 37 sampel (wilayah observasi berbeda seperti Sambeng, Kerek, Manyar, Tuban, dll).
- **Kolom Metadata (3 kolom):** `id`, `nama` (mahasiswa), `daerah` (wilayah kajian).
- **Kolom Fitur Deret Waktu (68 kolom):** 68 fitur TSFEL numerik.
- **Total Dimensi:** $37 \times 71$ kolom.

## 3. Migrasi Data Fitur ke Database Cloud PostgreSQL (Aiven)

Pada tahap ini, ketiga berkas hasil ekstraksi fitur (`ekstraksi_fitur_co.csv`, `ekstraksi_fitur_no2.csv`, dan `ekstraksi_fitur_so2.csv`) disimpan ke database PostgreSQL di cloud (Aiven). Hal ini bertujuan agar data dapat diakses secara terpusat oleh berbagai tools analisis, termasuk **KNIME Analytics Platform** melalui node *PostgreSQL Connector*.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

# 1. Load data time series hasil crawling Anda
df = pd.read_csv('ekstraksi_fitur_co.csv')

# 2. Setup koneksi ke PostgreSQL Aiven (Pastikan menggunakan postgresql:// dan port 15884)
# Jangan lupa ganti "password_anda" jika ini belum sesuai
AIVEN_DB_URL = "postgresql://avnadmin:password_anda@pg-9691fac-dedynurohim1-d965.i.aivencloud.com:15884/defaultdb?sslmode=require"

# Buat engine koneksi
engine = create_engine(AIVEN_DB_URL)

# 3. Masukkan data ke PostgreSQL
tabel_nama = 'data_clustering_co'
df.to_sql(tabel_nama, engine, if_exists='replace', index=False)

print(f"Berhasil! Data dari 'ekstraksi_fitur_co.csv' telah diupload ke tabel '{tabel_nama}' di Aiven PostgreSQL.")


Berhasil! Data dari 'ekstraksi_fitur_co.csv' telah diupload ke tabel 'data_clustering_co' di Aiven PostgreSQL.


In [2]:
import pandas as pd
from sqlalchemy import create_engine

# 1. Load data time series hasil crawling Anda
df = pd.read_csv('ekstraksi_fitur_no2.csv')

# 2. Setup koneksi ke PostgreSQL Aiven (Pastikan menggunakan postgresql:// dan port 15884)
# Jangan lupa ganti "password_anda" jika ini belum sesuai
AIVEN_DB_URL = "postgresql://avnadmin:password_anda@pg-9691fac-dedynurohim1-d965.i.aivencloud.com:15884/defaultdb?sslmode=require"

# Buat engine koneksi
engine = create_engine(AIVEN_DB_URL)

# 3. Masukkan data ke PostgreSQL
tabel_nama = 'data_clustering_no2'
df.to_sql(tabel_nama, engine, if_exists='replace', index=False)

print(f"Berhasil! Data dari 'ekstraksi_fitur_no2.csv' telah diupload ke tabel '{tabel_nama}' di Aiven PostgreSQL.")


Berhasil! Data dari 'ekstraksi_fitur_no2.csv' telah diupload ke tabel 'data_clustering_no2' di Aiven PostgreSQL.


In [3]:
import pandas as pd
from sqlalchemy import create_engine

# 1. Load data time series hasil crawling Anda
df = pd.read_csv('ekstraksi_fitur_so2.csv')

# 2. Setup koneksi ke PostgreSQL Aiven (Pastikan menggunakan postgresql:// dan port 15884)
# Jangan lupa ganti "password_anda" jika ini belum sesuai
AIVEN_DB_URL = "postgresql://avnadmin:password_anda@pg-9691fac-dedynurohim1-d965.i.aivencloud.com:15884/defaultdb?sslmode=require"

# Buat engine koneksi
engine = create_engine(AIVEN_DB_URL)

# 3. Masukkan data ke PostgreSQL
tabel_nama = 'data_clustering_so2'
df.to_sql(tabel_nama, engine, if_exists='replace', index=False)

print(f"Berhasil! Data dari 'ekstraksi_fitur_so2.csv' telah diupload ke tabel '{tabel_nama}' di Aiven PostgreSQL.")


Berhasil! Data dari 'ekstraksi_fitur_so2.csv' telah diupload ke tabel 'data_clustering_so2' di Aiven PostgreSQL.


## 4. Pembacaan dan Verifikasi Dimensi Data Hasil Ekstraksi Fitur

Pada tahap ini, kita memverifikasi bahwa ketiga tabel fitur polutan (CO, NO2, SO2) telah memiliki struktur dimensi yang lengkap dan seragam ($37 \text{ baris} \times 71 \text{ kolom}$), di mana 68 kolom di antaranya merupakan fitur numerik hasil ekstraksi TSFEL yang siap dianalisis.

In [5]:
import pandas as pd

# Membaca dan menampilkan berkas hasil ekstraksi fitur TSFEL
df_tsfel = pd.read_csv('ekstraksi_fitur_co.csv')

print(f"table : 'ekstraksi_fitur_co.csv' Dimensi data: {df_tsfel.shape[0]} baris x {df_tsfel.shape[1]} kolom fitur")
display(df_tsfel)


table : 'ekstraksi_fitur_co.csv' Dimensi data: 37 baris x 71 kolom fitur


,id,nama,daerah,abs_energy,auc,autocorr,average_power,calc_centroid,calc_max,calc_mean,...,spectral_spread,spectral_variation,spectrogram_mean_coeff,sum_abs_diff,wavelet_abs_mean,wavelet_energy,wavelet_entropy,wavelet_std,wavelet_var,zero_cross
0,3,Nurhabibatul Umah,"Kerek, Tuban",3.255516e-01,10.822536,3,8.919222e-04,185.172183,0.038546,0.029654,...,0.130288,0.554103,0.000015,0.704421,0.001815,0.008006,2.144113,0.007772,0.000069,0
1,4,Verdi Setyawan Ardiansyah Putra,Menganti Gresik,3.119461e-01,10.599365,3,8.546469e-04,183.614501,0.036701,0.029041,...,0.135135,0.592456,0.000015,0.709078,0.001715,0.007520,2.139121,0.007297,0.000060,0
2,5,Okan Syailendra wahyudi,"Manyar, Gresik",3.063319e-01,10.489406,3,8.415712e-04,180.902553,0.036488,0.028827,...,0.127778,0.661400,0.000014,0.623414,0.001676,0.007558,2.126252,0.007348,0.000061,0
3,6,Mochammad Zaenal Abidin,"Widang, Tuban",3.208447e-01,10.744960,3,8.790265e-04,184.819043,0.037897,0.029439,...,0.129537,0.474190,0.000013,0.698436,0.001767,0.007519,2.147071,0.007280,0.000060,0
4,7,M. Hendrik Purwanto,"Sreseh, Sampang",3.001284e-01,10.334234,4,8.313807e-04,178.342624,0.038664,0.028630,...,0.130322,0.684887,0.000017,0.611650,0.001698,0.007830,2.098579,0.007624,0.000068,0
5,8,Robbaniyah Umdatun Ni'mah,"Cerme, Gresik",2.315016e-01,7.741078,1,8.802342e-04,132.852591,0.038488,0.029437,...,0.147843,0.653533,0.000017,0.768211,0.002392,0.008140,2.139178,0.007740,0.000069,0
6,10,Ainur Raftuzzaki,Nunukan,2.786200e-01,10.016313,7,7.633425e-04,187.753748,0.034897,0.027451,...,0.123825,0.850024,0.000013,0.451485,0.001694,0.007603,2.083529,0.007400,0.000066,0
7,11,Magdalena Jovita Hatuopar,"Pilangkenceng, Madiun",3.191231e-01,10.705921,17,8.767118e-04,187.237450,0.037184,0.029420,...,0.128023,0.389429,0.000010,0.621369,0.001861,0.007592,2.124930,0.007341,0.000063,0
8,12,Rachelia Andini Tendean,"Tikala, Manado",2.337529e-01,9.101386,7,6.457263e-04,185.419585,0.033446,0.025156,...,0.131445,0.799799,0.000017,0.592317,0.001540,0.007593,2.112171,0.007422,0.000064,0
9,13,Kurnia Maulinda Sari,"Labang, Bangkalan",3.052118e-01,10.489433,3,8.361968e-04,180.420264,0.035705,0.028743,...,0.131499,0.615471,0.000012,0.622891,0.001695,0.007307,2.128615,0.007086,0.000058,0


In [7]:
import pandas as pd

# Membaca dan menampilkan berkas hasil ekstraksi fitur TSFEL
df_tsfel = pd.read_csv('ekstraksi_fitur_no2.csv')

print(f"table : 'ekstraksi_fitur_no2.csv' Dimensi data: {df_tsfel.shape[0]} baris x {df_tsfel.shape[1]} kolom fitur")
display(df_tsfel)


table : 'ekstraksi_fitur_no2.csv' Dimensi data: 37 baris x 71 kolom fitur


,id,nama,daerah,abs_energy,auc,autocorr,average_power,calc_centroid,calc_max,calc_mean,...,spectral_spread,spectral_variation,spectrogram_mean_coeff,sum_abs_diff,wavelet_abs_mean,wavelet_energy,wavelet_entropy,wavelet_std,wavelet_var,zero_cross
0,7,Okan Syailendra wahyudi,Manyar,1.933040e-06,0.023990,7,5.310551e-09,186.075084,0.000168,0.000066,...,0.147818,0.336845,1.288098e-09,0.005077,3.527742e-06,0.000043,2.147398,0.000043,1.940374e-09,0
1,10,Triswanti Jannatul Ma'wa,Kedungpring Lamongan,3.702058e-07,0.010905,8,1.014262e-09,194.285209,0.000059,0.000030,...,0.147428,0.319585,1.366945e-10,0.001801,1.783531e-06,0.000016,2.123802,0.000016,2.814972e-10,0
2,12,Laila Maghfiroh,"Kamal, Bangkalan",7.363589e-07,0.015059,11,2.017422e-09,166.039562,0.000086,0.000041,...,0.144113,0.282633,4.347435e-10,0.002744,1.433485e-06,0.000023,2.131037,0.000023,6.029032e-10,2
3,13,Dedy Nurohim,sambeng lamongan,3.854020e-07,0.011228,6,1.055896e-09,196.838820,0.000062,0.000031,...,0.156054,0.275151,1.258376e-10,0.002168,1.890919e-06,0.000013,2.154113,0.000013,1.879498e-10,0
4,14,Ahmad Syahrul Farihin,"Gresik Kota, Gresik",1.523980e-06,0.022417,3,4.175288e-09,184.039735,0.000124,0.000061,...,0.154967,0.685525,7.180607e-10,0.004380,2.944096e-06,0.000031,2.145496,0.000031,1.040073e-09,0
5,16,Rian Renaldy,"Waru, Pamekasan",1.123197e-07,0.005901,5,3.077252e-10,168.844890,0.000033,0.000016,...,0.148818,0.422768,7.722140e-11,0.001353,6.737109e-07,0.000010,2.142401,0.000010,1.046791e-10,6
6,18,Moh Rafie Nazar J,"Bangkalan Kota,Bangkalan",3.203641e-07,0.010196,3,8.849837e-10,184.972111,0.000057,0.000028,...,0.145301,0.729096,1.667061e-10,0.001859,1.326193e-06,0.000016,2.141427,0.000016,2.876259e-10,0
7,21,Ainur Raftuzzaki,Nunukan,2.363730e-08,0.002589,2,6.475970e-11,200.320300,0.000015,0.000007,...,0.155160,0.564399,2.510400e-11,0.000937,3.498942e-07,0.000005,2.172274,0.000005,3.113730e-11,16
8,25,Muhammad Zaidan Nabil Rafi,"Kamal, Bangkalan",1.883242e-06,0.023446,1,5.741593e-09,174.318724,0.000143,0.000072,...,0.163398,0.487746,9.295308e-10,0.006949,5.610511e-06,0.000034,2.169564,0.000034,1.189397e-09,0
9,27,Robbaniyah Umdatun Ni'mah,"Cerme, Gresik",7.517448e-07,0.012171,3,3.254306e-09,118.290125,0.000114,0.000053,...,0.153038,0.435391,6.793645e-10,0.003585,5.065708e-06,0.000028,2.180839,0.000027,7.646483e-10,0


In [8]:
import pandas as pd

# Membaca dan menampilkan berkas hasil ekstraksi fitur TSFEL
df_tsfel = pd.read_csv('ekstraksi_fitur_so2.csv')

print(f"table : 'ekstraksi_fitur_so2.csv' Dimensi data: {df_tsfel.shape[0]} baris x {df_tsfel.shape[1]} kolom fitur")
display(df_tsfel)


table : 'ekstraksi_fitur_so2.csv' Dimensi data: 37 baris x 71 kolom fitur


,id,nama,daerah,abs_energy,auc,autocorr,average_power,calc_centroid,calc_max,calc_mean,...,spectral_spread,spectral_variation,spectrogram_mean_coeff,sum_abs_diff,wavelet_abs_mean,wavelet_energy,wavelet_entropy,wavelet_std,wavelet_var,zero_cross
0,2,Nurhabibatul Umah,"Kerek, Tuban",4.612991e-06,0.028719,2,1.263833e-08,209.877355,0.000309,3.317126e-05,...,0.148927,0.209776,1.899262e-08,0.027288,8.060801e-07,0.000153,2.160229,0.000153,2.464593e-08,93
1,3,Verdi Setyawan Ardiansyah Putra,Menganti Gresik,1.091218e-05,0.042161,2,2.989639e-08,172.148705,0.000461,8.686927e-05,...,0.154508,0.292423,4.194986e-08,0.038801,6.563252e-06,0.000204,2.157516,0.000204,4.519175e-08,100
2,4,Okan Syailendra wahyudi,"Manyar, Gresik",2.477941e-05,0.063998,1,6.807531e-08,204.258058,0.000793,1.113751e-04,...,0.153556,0.278411,9.647455e-08,0.064800,6.723475e-06,0.000310,2.168640,0.000310,1.002558e-07,98
3,5,Mochammad Zaenal Abidin,"Widang, Tuban",3.845977e-06,0.026307,2,1.053692e-08,186.794715,0.000274,2.567220e-05,...,0.147246,0.172626,1.702950e-08,0.025822,3.133062e-06,0.000142,2.153694,0.000142,2.183313e-08,107
4,6,M. Hendrik Purwanto,"Sreseh, Sampang",3.383721e-06,0.022927,2,9.373189e-09,210.032798,0.000293,2.111224e-05,...,0.147471,0.220521,1.625267e-08,0.026505,1.106266e-06,0.000132,2.167505,0.000132,1.856150e-08,108
5,7,Robbaniyah Umdatun Ni'mah,"Cerme, Gresik",3.198463e-07,0.006406,1,8.811193e-10,205.784079,0.000086,1.175118e-05,...,0.154676,0.233806,1.265747e-09,0.010152,9.849967e-07,0.000030,2.191516,0.000030,8.844974e-10,168
6,9,Ainur Raftuzzaki,Nunukan,7.229314e-07,0.010306,1,1.980634e-09,180.522256,0.000107,-2.049584e-06,...,0.145901,0.204832,3.709163e-09,0.014261,5.053866e-07,0.000056,2.185740,0.000056,3.197403e-09,122
7,10,Magdalena Jovita Hatuopar,"Pilangkenceng, Madiun",3.986450e-06,0.025926,1,1.095178e-08,183.041373,0.000304,6.493568e-05,...,0.151660,0.318823,1.162739e-08,0.026299,5.375518e-06,0.000093,2.188757,0.000093,8.791744e-09,95
8,11,Rachelia Andini Tendean,"Tikala, Manado",3.801074e-06,0.025848,2,1.050020e-08,199.981451,0.000304,-1.006920e-05,...,0.144522,0.258207,1.832777e-08,0.022556,8.380800e-07,0.000153,2.150468,0.000153,2.499851e-08,81
9,12,Marshella Aulia Putri,"Kec. Kalianget, Sumenep",3.993830e-06,0.018879,1,1.685160e-08,115.260448,0.000328,1.845555e-05,...,0.150961,0.243976,3.262437e-08,0.029970,4.297109e-06,0.000144,2.192455,0.000144,2.089430e-08,110


## 5. Metodologi Reduksi Dimensi Menggunakan PCA (Principal Component Analysis)

### Mengapa Reduksi Dimensi Dilakukan Menjadi 37 Komponen?
1. **Batas Teori Aljabar Linier (*Matrix Rank*):**
   - Matriks data fitur kita memiliki ukuran $N = 37$ baris (sampel mahasiswa) dan $P = 68$ kolom (fitur).
   - Secara matematis, rank maksimal dari matriks kovarians data adalah:
     $$\text{Rank Maksimum} = \min(N, P) = \min(37, 68) = 37$$
   - Artinya, meskipun secara nominal terdapat 68 fitur, seluruh titik data 37 sampel tersebut sesungguhnya berada dalam subruang linier berdimensi maksimal 37.
2. **Preservasi 100% Variansi (*Cumulative Explained Variance*):**
   - Dengan mereduksi dimensi dari 68 fitur menjadi **37 komponen utama (PC1 s.d. PC37)**, model mempertahankan **100.00% informasi variansi data asli** tanpa ada informasi yang hilang.
3. **Penyederhanaan ke Ruang 2D (PC1 & PC2):**
   - Berdasarkan hasil perhitungan variansi:
     - **Polutan CO:** PC1 dan PC2 merangkum **77.63%** variansi data.
     - **Polutan NO2:** PC1 dan PC2 merangkum **68.58%** variansi data.
     - **Polutan SO2:** PC1 dan PC2 merangkum **77.46%** variansi data.
   - Hal ini membuktikan bahwa proyeksi 2D (PC1 vs PC2) sudah sangat representatif untuk visualisasi *Scatter Plot* tanpa mengalami *curse of dimensionality*.

---

## 6. Analisis Penentuan Jumlah Cluster Terbaik pada K-Means Clustering

Untuk menentukan jumlah klaster optimal ($k$), dua metode evaluasi kuantitatif digunakan:
1. **Metode Elbow (Inertia / WCSS - Within-Cluster Sum of Squares):**
   Mengukur derajat kepadatan internal klaster. Titik "siku" (*elbow*) menunjukkan titik di mana penambahan klaster baru tidak lagi memberikan penurunan inertia yang signifikan.
2. **Metode Silhouette Score:**
   Mengukur seberapa mirip suatu objek dengan klasternya sendiri dibandingkan dengan klaster tetangga. Rentang nilai antara $-1$ hingga $+1$. Nilai di atas $0.70$ mengindikasikan pemisahan klaster yang sangat kuat (*strong clustering structure*).

### Rangkuman Hasil Evaluasi Klaster ($k=2$ s.d. $k=7$):

| Polutan | Nilai $k$ | Silhouette Score | Davies-Bouldin Index | Inertia (WCSS) | Status Evaluasi |
| :--- | :---: | :---: | :---: | :---: | :--- |
| **CO** | **$k=2$** | **0.7955** | **0.1247** | 1123.80 | **Optimal Tertinggi (Sangat Kuat)** |
| | **$k=3$** | **0.6970** | **0.1591** | 660.21 | **Alternatif Terbaik (3 Kategori)** |
| | $k=4$ | 0.3425 | 0.6945 | 509.37 | Kualitas turun tajam |
| | $k=5$ | 0.1845 | 0.9275 | 413.89 | Terjadi over-clustering |
| **NO2** | **$k=2$** | **0.7455** | **0.1631** | 1326.58 | **Optimal Tertinggi (Sangat Kuat)** |
| | $k=3$ | 0.1907 | 1.1257 | 1001.28 | Kualitas rendah |
| | $k=4$ | 0.2045 | 0.9997 | 811.80 | Klaster tumpang tindih |
| **SO2** | **$k=2$** | **0.7894** | **0.1279** | 1210.46 | **Optimal Tertinggi (Sangat Kuat)** |
| | **$k=3$** | **0.7085** | **0.1513** | 691.64 | **Alternatif Terbaik (3 Kategori)** |
| | $k=4$ | 0.3921 | 0.5334 | 525.29 | Kualitas turun tajam |

---

### Jawaban Pertanyaan Tugas:

#### 1. Dari hasil reduksi dimensi PCA 1 – PCA 37 berapa baiknya cluster?
- **Jumlah cluster terbaik secara matematis adalah $k = 2$** dengan nilai Silhouette Score yang sangat tinggi (**CO: 0.7955, NO2: 0.7455, SO2: 0.7894**). Pemisahan ini membagi wilayah secara tegas menjadi 2 kelompok profil emisi (misalnya wilayah dengan aktivitas polusi tinggi vs rendah).
- **Alternatif praktis terbaik untuk CO dan SO2 adalah $k = 3$** (Silhouette $\approx 0.70$), yang membagi kualitas lingkungan menjadi 3 tingkatan logis: **Rendah, Sedang, dan Tinggi**.

#### 2. Berapa baiknya data ini di-cluster?
- Data ini **sangat baik untuk di-cluster pada $k=2$ atau $k=3$**, terbukti dengan Silhouette Score di atas ambang $0.70$ (*strong structure*). 
- Penggunaan $k \ge 4$ **tidak direkomendasikan** karena skor Silhouette anjlok ke bawah $0.40$, menandakan terjadinya fragmentasi berlebih (*over-clustering*) mengingat ukuran sampel sebanyak 37 data.

#### 3. Bagaimana hasilnya jika dilakukan pada 68 fitur asli secara langsung?
- Nilai Silhouette Score dan inertia pada 68 fitur asli **identik** dengan hasil pada 37 komponen PCA. Hal ini terjadi karena reduksi PCA ke 37 komponen merupakan transformasi ortogonal penuh yang mempertahankan 100% variansi dan jarak Euclidean antar-data.
- **Keunggulan reduksi dimensi PCA:** Menyediakan sumbu komponen utama (PC1 & PC2) yang dapat langsung dipetakan ke dalam **Scatter Plot 2D** pada KNIME maupun Python tanpa kehilangan pola esensial data.

---

## 7. Panduan Implementasi pada Tools KNIME Analytics Platform

Sesuai diagram node workflow KNIME (*PostgreSQL Connector &rarr; DB Table Selector &rarr; DB Reader &rarr; PCA &rarr; k-Means &rarr; Scatter Plot*), berikut konfigurasi tiap node:

1. **PostgreSQL Connector:**
   - Masukkan *Hostname* Aiven, *Port* `15884`, *Database* `defaultdb`, serta *Username* dan *Password*.
2. **DB Table Selector:**
   - Pilih tabel target (`data_clustering_co`, `data_clustering_no2`, atau `data_clustering_so2`).
3. **DB Reader:**
   - Eksekusi pembacaan data ke dalam memori tabel KNIME.
4. **PCA Node:**
   - Pindahkan kolom non-numerik (`id`, `nama`, `daerah`) ke kotak *Exclude*.
   - Masukkan seluruh 68 kolom fitur numerik ke kotak *Include*.
   - Pada pengaturan dimensi target, pilih **37 Dimensions** (atau pertahankan *100% Variance*).
5. **k-Means Node:**
   - Masukkan parameter **$k = 2$** (atau **$k = 3$**).
   - Centang opsi *Normalize input variables* agar skala fitur seragam.
6. **Scatter Plot Node:**
   - Konfigurasikan sumbu horizontal $X = \text{PCA 1}$ dan sumbu vertikal $Y = \text{PCA 2}$.
   - Atur warna titik berdasarkan kolom hasil pengelompokan **Cluster**.